### PR TITLE
Recursively remove dependencies when using yay -Yc

### PR DESCRIPTION
**Pull request by ratorx**
Fix https://github.com/Jguer/yay/issues/268

* Recursively removes unneeded dependencies
* Preserves optional dependencies by default (overridden with -cc)
* Does not list package removals, just passes them into pacman and relies on pacman confirmation dialog

### CODE REVIEW

**Comment by Jguer** [comment]
Any easy packages to install and uninstall to test functionality?

**Comment by Morganamilo** [comment]
I'm trying with git, has a lot of optional stuff.

**Comment by Morganamilo** [comment]
Something doesn't quite work here. Yay should persevere optional dependencies for stuff we are keeping. But if we are removing `foo` I want the un-needed  optional deps for `foo` uninstalled also.

For example I have everything installed as a dependency, not one package on my system is listed as explicit. Yet doing `yay -c` It wants to keep the opt depends for git installed, even though git is going to be removed by `yay -c`

Easy way to reproduce:
```
mkdir ~/db
yay -Sy -b ~/db
yay -S -b ~/db --asdeps tk perl-libwww perl-term-readkey perl-mime-tools perl-net-smtp-ssl perl-authen-sasl perl-mediawiki-api perl-datetime-format-iso8601 perl-lwp-protocol-https perl-cgi python2 subversion gnome-keyring git
yay -c
```

Also I would rebase out the merge commits, they're just clutter really.

**Comment by ratorx** [comment]
I had a look into it. It seems to be a problem with `pacman -Rs` rather than any code in yay.

I tested this by doing:
```
pacman -S --asdeps git perl-term-readkey # Just one of the optional dependencies to be minimal
```

When `pacman -Rs git`, only direct dependencies are removed (in my case `perl-error`)
However, this leaves `perl-term-readkey` unneeded dependencies behind (found by both `yay -c` and `pacman -Qdt`).

A potential fix would be to implement a recursive walk in yay (what I had originally), instead of relying on `pacman -Rs`.

I found another problem however - dependency cycles are not resolved.

As an example, if I do `pacman -S --asdeps neovim python-neovim` and then try to remove unnecessary packages (using either `pacman -Rs $(pacman -Qtdq)` or `yay -c`), neither of them picks these packages up. I would have to think more about how to fix this, but if you have any ideas, let me know.

**Comment by Morganamilo** [comment]
Doing it recessive is probably the best way to deal with it. Not sure on the exact implementation though.

As for circular stuff. Maybe keep a cache of your current tree, then as you work your way up if you find something that's in that cache assume it's not needed. No idea if it will work, just an idea.

**Comment by Morganamilo** [comment]
As for implementation my current thought is: assume everything is to be removed, iterate over explicit packages and mark everything in there to be kept, remove what is not to be kept.

Also as well as keeping optional deps, I think we should keep makedeps for AUR packages.

**Comment by ratorx** [comment]
That seems very doable. It should deal with circular and optional dependencies as well.

Can you point me towards any existing code that gets makedeps from the AUR? My current understanding is that it can't be accessed through libalpm?

**Comment by Morganamilo** [comment]
You can do it in the latest git for alpm :(

Infact I was using the git version while developing go-alpm and got confused when the functions didn't work, hence:
image

You'll have to use `aurInfo()` to get the makedeps.

**Comment by Morganamilo** [comment]
Any reason you closed this?

**Comment by ratorx** [comment]
It was accidental, I undid some of the merge commits and rebased, but I'll leave it closed for now. I'll reopen when I have a fixed version.

**Comment by Morganamilo** [comment]
Alright, if you're starting over I would recommend `git pull --rebase` over `git pull`.

**Comment by ratorx** [comment]
Will do. I've got the basic thing recursion package marking working, just need to add support for AUR makedeps.

**Comment by ratorx** [comment]
I had a look into getting makedeps for AUR packages. Considering that using aurinfo involves making web requests and adds a fair bit of complexity to determining hanging packages, do you think its worth it to determine and keep makedeps?

The alternative would be to wait for stable libalpm to be updated so that the bindings in go-alpm work and then integrate that.

**Comment by Morganamilo** [comment]
I don't really mind either way.

I wouldn't say using `aurInfo()` adds that much complexity though, the function is pretty easy to use. The internet access is the more annoying thing, you wouldn't expect Yay to need that for removing packages.

Pacman v5.0.2 was released `Sat Jun 3 22:17:02 2017 ` so who knows when the next update will be.

**Comment by ratorx** [comment]
I've redone it now, and tested it a bit - it seems to work!

**Comment by Morganamilo** [comment]
Looking at the vendor update it looks like you edited the file manually. Instead you should run `dep ensure -update` to update the deps.

**Comment by Morganamilo** [comment]
Also run `make test` and to lint `go fmt`.

**Comment by ratorx** [comment]
Ahh sorry, didn't realise. Done both now.

**Comment by Morganamilo** [comment]
Tested the code a little and it seems like it works alright.

This is false though:
```
	// Assumption - multiple installed packages don't provide the same dependency
```

So this would probably fail on something like java.

**Comment by ratorx** [comment]
Ahh alright, I can work around that assumption, just wasn't sure whether I needed to when I wrote the code.

File: query.go [hunk 1]
```diff
@@ -300,7 +299,13 @@ func hangingPackages(removeOptional bool) (hanging []string, err error) {
 		}
 
 		pkg.Provides().ForEach(func(dep alpm.Depend) error {
-			provides[dep.Name] = pkg.Name()
```
Comment (Morganamilo) [inline query.go]: There's a function called `addMapStringSet` that does all this for you ;)

feel free to create `addMapStringString` as well or find a generic way.

Comment (ratorx) [reply query.go]: I was not aware of `addMapStringSet` and I'll rewrite my code.

I'm not too sure what you mean by `addMapStringString`, or which part of my code you think I should use it in.

Comment (Morganamilo) [reply query.go]: Yeah I wasn't thinking when I said that. Was thinking about `map[stiring]string` but obviously you can set that directly.

**Comment by Jguer** [comment]
Let's leave this patch for the next release cycle as it's still undergoing changes, you're the closest on it @Morganamilo , is that ok with you?

**Comment by Morganamilo** [comment]
That was my plan already. Would like my two PRs merged though.

**Comment by Morganamilo** [comment]
I gave this a test again and it seems to work fully now so nice job. I still need to make sure though and give the code a proper read.

**Comment by Morganamilo** [comment]
Also is there any reason you're not using ComputRequiredBy/ComputreOptionalFor anymore?.

**Comment by ratorx** [comment]
In order to deal with cyclic dependencies, I switched strategy, so I now need to get the depends and optdepends for a given package, rather than the packages which depend/optionally depend on it.

**Comment by Morganamilo** [comment]
I guess we had opposite idea's then. You want to mark the explicit packages, find the stuff it depends on and loop through the package list to find them. My idea was to mark the explicit packages, iterate the package list calling the Compute*() functions on them and see which ones are required by the explicit packages.

As a bonus the Compute*() functions take versioned depends into account which is nice to support even if it's not really needed.

**Comment by Morganamilo** [comment]
This worked when I did some limited testing last time. New release is out and a new window is open so I'll give this some testing now and merge it. In the meantime conflicts need to be resolved.

**Comment by Morganamilo** [comment]
Just reminding you that you need to resolve these conflicts. If you don't know how GitHub has their own tutorial for it. https://help.github.com/articles/resolving-a-merge-conflict-on-github/

**Comment by Morganamilo** [comment]
Could you squash/rebase this? Makes things neater but not that important.

**Pull request merged**